sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/PruneHiveTablePartitions.scala

```diff
@@ -74,6 +74,32 @@ private[sql] class PruneHiveTablePartitions(session: SparkSession)
         0L
       }
     }
+    if (sizeOfPartitions.forall(_ > 0) && prunedPartitions.forall(_.stats.isDefined)) {
+      val rowCountOfPartitions = prunedPartitions.map { partition =>
+        val rowCount = partition.stats.get.rowCount.map(_.toLong)
+        if (rowCount.isDefined && rowCount.get > 0L) {
+          rowCount.get
+        } else {
+          0L
+        }
+      }
+      if (rowCountOfPartitions.forall(_ > 0)) {
+        val stats = relation.stats
+        val oldRowCount = stats.rowCount
+        val newRowCount = BigInt(rowCountOfPartitions.sum)
+        var colStats = Map.empty[String, CatalogColumnStat]
+        if (oldRowCount.isDefined) {
+          stats.attributeStats.map(_._2.updateCountStats(
+            oldNumRows = oldRowCount.get, newNumRows = newRowCount))
+          colStats = stats.attributeStats.map{ case (attr, colStat) =>
+            (attr.name, colStat.toCatalogColumnStat(attr.name, attr.dataType))}
+        }
+        return relation.tableMeta.copy(stats = Some(CatalogStatistics(
+          sizeInBytes = BigInt(sizeOfPartitions.sum),
+          rowCount = Option(newRowCount),
+          colStats = colStats)))
+      }
+    }
     if (sizeOfPartitions.forall(_ > 0)) {
```

Comment:
I think we can put the new code inside this `if` branch. The idea is to use the accurate row count from partitions directly instead of estimating it.
```
val rowCount = if (prunedPartitions.forall(_.stats.flatMap(_.rowCount).filter(_ > 0).isDefined)) {
  ...
} else {
  ...
}
```

Reply:
@cloud-fan Thanks for your advice. I have modified the code and submitted it again